### ISSUE TITLE
[JIT] Profiling executor is not fusing Dropout while legacy executor does

### ISSUE THREAD

**Issue by kevinstephano** [pytorch/pytorch#45519]
## 🐛 Bug

For a simple fusion of RELU + Dropout, the profiling executor does not make the fusion and breaks the operation into its elemental operations.  You can see this from the kernels profiled below.

```
7.07477s  35.616us          (20480 1 1)        (64 1 1)        21        0B        0B  Tesla V100-DGXS         1         7  _ZN2at6native29vectorized_elementwise_kernelILi4EZNS0_21threshold_kernel_implIN3c104HalfEEEvRNS_14TensorIteratorET_S7_EUlS4_S4_E_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_ [495]
7.07844s  22.272us            (640 1 1)       (256 1 1)        46        0B        0B  Tesla V100-DGXS         1         7  _ZN2at6native89_GLOBAL__N__65_tmpxft_000048e2_00000000_8_DistributionUniform_compute_75_cpp1_ii_8b4e27ca43distribution_elementwise_grid_stride_kernelIfLi4EZNS0_9templates4cuda21uniform_and_transformIN3c104HalfEfLm4EPNS_17CUDAGeneratorImplEZZZNS4_14uniform_kernelIS9_EEvRNS_14Tens
orIteratorEddT_ENKUlvE_clEvENKUlvE1_clEvEUlfE_EEvSC_T2_T3_EUlP24curandStatePhilox4_32_10E0_ZNS1_27distribution_nullary_kernelIS7_fLi4ES9_SL_SG_EEvSC_SH_RKSI_T4_EUlifE_EEviSt4pairImmET1_SH_ [510]
7.08192s  35.584us          (20480 1 1)        (64 1 1)        17        0B        0B  Tesla V100-DGXS         1         7  _ZN2at6native29vectorized_elementwise_kernelILi4ENS0_13BUnaryFunctorIZZZNS0_14lt_kernel_cudaERNS_14TensorIteratorEENKUlvE_clEvENKUlvE6_clEvEUlN3c104HalfES8_E_EENS_6detail5ArrayIPcLi2EEEEEviT0_T1_ [523]
7.08466s  45.440us          (20480 1 1)        (64 1 1)        32        0B        0B  Tesla V100-DGXS         1         7  _ZN2at6native27unrolled_elementwise_kernelIZZZNS0_21copy_device_to_deviceERNS_14TensorIteratorEbENKUlvE0_clEvENKUlvE8_clEvEUlN3c104HalfEE_NS_6detail5ArrayIPcLi2EEE23TrivialOffsetCalculatorILi1EjESE_NS0_6memory12LoadWithCastILi1EEENSF_13StoreWithCastEEEviT_T0_T1_T2_T3_T4_ [537]
7.08536s  41.152us          (20480 1 1)        (64 1 1)        19        0B        0B  Tesla V100-DGXS         1         7  _ZN2at6native29vectorized_elementwise_kernelILi4EZZZNS0_15mul_kernel_cudaERNS_14TensorIteratorEENKUlvE0_clEvENKUlvE8_clEvEUlN3c104HalfES7_E_NS_6detail5ArrayIPcLi3EEEEEviT0_T1_ [548]
7.08680s  36.000us          (20480 1 1)        (64 1 1)        17        0B        0B  Tesla V100-DGXS         1         7  _ZN2at6native29vectorized_elementwise_kernelILi4EZZZNS0_15mul_kernel_cudaERNS_14TensorIteratorEENKUlvE_clEvENKUlvE3_clEvEUlN3c104HalfEE_NS_6detail5ArrayIPcLi2EEEEEviT0_T1_ [557]
```
With the legacy executor the following fusion occurs:

```
7.66330s  51.232us           (1280 1 1)       (128 1 1)        30        0B        0B  Tesla V100-DGXS         1         7  kernel_0 [488]
```

## To Reproduce

The script to reproduce is as follows.  Not the `--bad-jit` option turns on the profiling executor.

```
import torch
import torch.nn as nn
import torch.nn.functional as F
import argparse

parser = argparse.ArgumentParser(description='Relu-Dropout JIT Test')
parser.add_argument('--bad-jit', action='store_true', help='Use new executor that cause Relu not to fuse with dropout.')
args = parser.parse_args()

assert hasattr(torch._C, '_jit_set_profiling_executor'), "Old JIT behavior doesn't exist!"
if not args.bad_jit :
    torch._C._jit_set_profiling_executor(False)
    torch._C._jit_set_profiling_mode(False)
else :
    torch._C._jit_set_profiling_executor(True)
    torch._C._jit_set_profiling_mode(True)

@torch.jit.script
def jit_relu_dropout(x, prob) :
    # type: (Tensor, float) -> Tensor
    out = F.threshold(x, 0., 0.)
    out = F.dropout(out, p=prob, training=True)
    return out 
                                                                                    
inputs = torch.randn(5120, 1024, dtype=torch.float16, device=torch.device("cuda")).requires_grad_(True)

for x in range(0, 5) :
    outputs = jit_relu_dropout(inputs, 0.1)
```

## Environment

PyTorch version: 1.7.0a0+9093ffa
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.14.0

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-DGXS-16GB
GPU 1: Tesla V100-DGXS-16GB
GPU 2: Tesla V100-DGXS-16GB
GPU 3: Tesla V100-DGXS-16GB

Nvidia driver version: 440.33.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.2
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.2
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.2
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.2
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] numpy==1.19.1
[pip] torch==1.7.0a0
[conda] magma-cuda102             2.5.2                         1    pytorch
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] nomkl                     3.0                           0  
[conda] numpy                     1.19.1           py38h30dfecb_0  
[conda] numpy-base                1.19.1           py38h75fe3a5_0  
[conda] torch                     1.7.0a0                  pypi_0    pypi

## Additional context

cc @gmagogsfm

**Comment by Lilyjjo** [comment]
@Krovatkin is this in the scope of PE/NNC?

**Comment by Krovatkin** [comment]
Yes, we will tackle this in 1.8.

**Comment by kevinstephano** [comment]
Just to motivate the problem, in case you don't have the context, the dropout operator is heavily used in Transformer/Bert NLP networks and being able to fuse to dropout is an important perf opportunity in those networks.

**Comment by jjsjann123** [comment]
Summarize the few problem we are observing:

1. Branching is the source of all evil:

Look at dropout expansion: https://github.com/pytorch/pytorch/blob/8b13ab9370df613aa05966b5fe79962adaac4ca1/torch/csrc/jit/runtime/symbolic_script.cpp#L1083-L1111
as well as torch.nn.Linear:
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1663-L1671

Branching is blocking fusion of these with consecutive operations.

2. Constant folding doesn't work for profiled information with non-BailOut PE.

Take another look at the dropout example vvv, `use_cuda` cannot be constant folded, because we are not extracting/guarding profiled information when we are not using BailOut node:
```
            use_cuda = input.is_cuda

            # ....

            if train:
                if use_cuda:
                    mask = torch.rand_like(input, memory_format=1) < p1m
                    res = mask.type_as(input) * input * (1./p1m)
                else:
                    mask = torch.empty_like(input, memory_format=1)
                    mask.bernoulli_(p1m)
                    res = mask * input / p1m
            else:
                p1m = 1.
                res = input
                mask = torch.empty_like(input, memory_format=1)
```

The same thing applies to `torch.nn.Linear` module, where rank of input tensor can't be constant folded to eliminate branchings.

3. A very minor issue, there's a bug in `graph_fuser.cpp` that we forgot to pass `strict_fuser_check`. So right now there's no way to turn off legacy fuser on dropout. I have a small PR already: https://github.com/pytorch/pytorch/pull/47221

**Comment by nikithamalgifb** [comment]
@penguinwu , does this need triage in JIT? Looks like you already moved it to progress. I will remove the jit triage label for now . Feel free to re-add if necessary.

**Comment by kevinstephano** [comment]
I think you might be able to close this.  I think we have a general solution to Dropout fusion problems to upstream.

**Comment by jjsjann123** [comment]
Thanks to @kevinstephano for chime in. I'm closing this issue.
